kvserver: [wip] don't transfer leases to overloaded stores

kvoli opened the pull request:
Previously, it was possible for a lease transfer to occur towards a
store which is IO overloaded. This commit adds a check among valid,
candidate replicas that excludes replicas which are on store that exceed
the io overload threshold and are greater than the 1.1 * mean io
overload.

In addition, lease transfers will now only consider the average load
(lease count, qps, cpu, locality load) of valid replica stores when comparing
candidate replicas as potential targets.

Release note: None

kvoli committed:
kvserver: replace read amp with io threshold

L0-sublevels used to be checked as a gauge of store health in order to
exclude unhealthy stores from being allocation targets. In #83720,
`IOThreshold` was included into gossip. `IOThreshold` includes
the L0-sublevels, L0-filecount and may be used for additional
measures of store health, making it more informative than looking at
`L0-sublevels` alone.

This commit stops gossiping the L0-sublevels and replaces uses of
L0-sublevels in the allocator and storepool with `IOThreshold`.

The prior cluster settings which controlled how store health was
considered in allocation are now deprecated:

`kv.allocator.L0_sublevels_threshold`
`kv.allocator.L0_sublevels_threshold_enforce`

The new cluster settings which perform an identical function:

`kv.allocator.io_overload_threshold`
Which has the same concept as `L0-sublevels_threshold`, however is
adjusted for `IOThreshold` where a value above 1 indicates overload. The
default cluster setting value is set to `0.8`, which is intentional to
prevent allocation before overload occurs.

`kv.allocator.io_overload_threshold_enforce`
Which is identical to the `L0_sublevels_threshold_enforce`.

Resolves: #85084

Release note (ops change): Deprecate cluster settings
`kv.allocator.l0_sublevels_threshold` and
`kv.allocator.L0_sublevels_threshold_enforce` and introduce
`kv.allocator.io_overload_threshold` and
`kv.allocator.io_overload_threshold_enforce`. The enforcement cluster
setting is unchanged, however renamed. The threshold cluster setting is
adapted to `IOThreshold`, where a value greater or equal to 1 indicates
IO overload. The default is set to 0.8 to prevent allocation prior to
overload occurring.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96994)
<!-- Reviewable:end -->

kvoli committed:
storepool: [wip] consider stores suspect on join

Previously, stores would not be considered suspect when joining the
cluster. This patch updates logic so that stores are now considered
suspect on joining the cluster (rejoining or startup).

Release note (ops change): Stores will now be considered suspect when
joining the cluster.

kvoli commented:
closing in favor of #97587
